src/storage.rs

```diff
@@ -56,6 +56,45 @@ impl RaftState {
     }
 }
 
+/// Records the context of the caller who calls entries() of Storage trait.
+#[derive(Debug)]
+pub struct GetEntriesContext(pub(crate) GetEntriesFor);
+
+impl GetEntriesContext {
+    /// Used for callers out of raft. Caller can customize if it supports async.
+    pub fn empty(can_async: bool) -> Self {
+        GetEntriesContext(GetEntriesFor::Empty(can_async))
+    }
+
+    /// Check if the caller's context support fetching entries asynchrouously.
+    pub fn can_async(&self) -> bool {
+        match self.0 {
+            GetEntriesFor::SendAppend { .. } => true,
+            GetEntriesFor::Empty(can_async) => can_async,
+            _ => false,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum GetEntriesFor {
+    // for sending entries to followers
+    SendAppend {
+        /// the peer id to which the entries are going to send
+        to: u64,
+        /// whether to exhaust all the entries
+        aggressively: bool,
+    },
+    // for getting committed entries in a ready
+    GenReady,
+    // for getting entries to check pending conf when transferring leader
+    TransferLeader,
+    // for getting entries to check pending conf when forwarding commit index by vote messages
+    CommitByVote,
+    // It's not called by the raft itself
+    Empty(bool),
```

Comment:
Do we need `Empty(true)`? it is a noop for `on_entries_fetched`

Reply:
Yes, TiKV needs it to test the process of async fetch